### ISSUE TITLE
Wishlist: is_dag should return a cycle, when there is one

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1251]
This is a feature suggestion.

Currently, `igraph_is_dag()` only returns whether the graph is a DAG.

If it is a directed graph with a cycle (i.e. not a DAG), `is_dag` should optionally return proof of this, i.e. a directed cycle.

Suggested prototype:

```c
igraph_is_dag(const igraph_t* graph, igraph_bool_t *res, igraph_vector_t *cycle);
```

If `cycle != NULL`, the function should return the ordered list of edges (edge indices) participating in the detected cycle.

Any thoughts?

**Comment by ntamas** [comment]
How about adding `igraph_find_cycle(const igraph_t* graph, igraph_vector_int_t* cycle)` that would return the first cycle it finds in `cycle` or return an empty vector instead if there are no cycles?

This way we won't have to break the API (although we would be breaking it anyway for the next version), we could have a simple cycle-finder function that works even for undirected graphs (if we implement it that way) and we could also re-implement `igraph_is_dag()` as `igraph_is_directed() && igraph_find_cycle()`.

**Comment by szhorvat** [comment]
Yes, that's probably a better idea, and would naturally generalize to undirected graphs as well. I would support this naming.

----

One thing to consider is that users may expect such a function to be able to find more than one cycle. Mathematica has a [FindCycle](https://reference.wolfram.com/language/ref/FindCycle.html) function which can find "up to `k` cycles", based on the `k` parameter. 

I would not want to block the implementation for finding a _single_ cycle until we have something that can find multiple cycles (which would be more work), but it's good to think about function naming in advance.

----

Some more thoughts:

 - This request was inspired by SageMath's approach of many graph theory functions returning am optional "certificate", i.e. some sort of object whose mere existence demonstrates that the solution is valid. E.g. the existence of a cycle shows that the graph is not acyclic.

 - If we ever implement a system to enumerate _all_ cycles, the iterator paradigm may be a better approach anyway because of the exponential number of cycles. I.e. construct iterator, then generate cycles _one by one_ until there are no more or until the user decides that they've seen enough cycles.

Related: https://github.com/igraph/igraph/issues/379

**Comment by szhorvat** [comment]
@ntamas Do you prefer `igraph_vector_int_t *cycle` or `igraph_vector_t *cycle`? While the first makes more sense, currently the standard for edge indices and vertex indices is the 2nd. I would do the 2nd because then the result can be passed to other functions more easily.

Also, I propose representing the cycle through its edges, not its vertices, as this makes more sense in multigraphs.

Sounds good?

**Comment by ntamas** [comment]
Tricky question :) Historically, functions that take or return vertex or edge indices used `igraph_vector_t`, because we haven't had `igraph_vector_int_t` for a long while (we only had `igraph_vector_t`, nothing else). Later on, we have started adding other vector types. At that point, we could have switched to using `igraph_vector_long_t` for vertex and edge indices because it has caused confusion several times for inexperienced users that the elements of an `igraph_vector_t` are doubles, even if they store integers only. However, we did not make the switch, and it's becoming increasingly harder to do it as the API surface of the library becomes larger.

I think that for the time being we could stick to the "majority rule" and use `igraph_vector_t`.

**Comment by vtraag** [comment]
Perhaps we should at one point switch to `igraph_vector_int_t`? Or perhaps even `igraph_vector_size_t` since they should be non-negative numbers anyhow? I believe  `igraph_size_t` as a type does not yet exist, but it would make sense to introduce it I think. Perhaps we should open a separate issue for that?

**Comment by szhorvat** [comment]
At some point in the future, this should definitely be tackled.

It's a high-risk change though, and it's also a lot of work. In the spirit of "if it works, don't fix it" I would leave it for much later, and only after we have a better test suite (and maybe benchmark so we can see the performance improvements, if there are any). Perhaps for v1.0 (or the version prior to that) since it's a large API-breaking change.

While all the `double`s annoy me too, I think that for 0.9 we can allocate our limited developer time to projects that have clearer benefits.

> Perhaps we should open a separate issue for that?

This one looks like it, but feel free to open a new one of you judge that it's useful.

https://github.com/igraph/igraph/issues/556

**Comment by szhorvat** [comment]
`igraph_find_cycle()` now does this.
